stability-tests/daa/daa_test.go

```diff
@@ -0,0 +1,340 @@
+						int64(math.Pow(float64(toHashNanoseconds-fromHashNanoseconds), timeElapsedFraction))
+				}
+
+				// 5 minute cooldown
+				return toHashNanoseconds
+			},
+		},
+	}
+
+	for _, test := range tests {
+		runDAATest(t, test.name, test.runDuration, test.targetHashNanosecondsFunction)
+	}
+}
+
+func measureMachineHashNanoseconds(t *testing.T) int64 {
+	t.Logf("Measuring machine hash rate")
+	defer t.Logf("Finished measuring machine hash rate")
+
+	genesisBlock := dagconfig.DevnetParams.GenesisBlock
+	targetDifficulty := difficulty.CompactToBig(genesisBlock.Header.Bits())
+	headerForMining := genesisBlock.Header.ToMutable()
+
+	machineHashesPerSecondMeasurementDuration := 10 * time.Second
+	hashes := int64(0)
+	runForDuration(machineHashesPerSecondMeasurementDuration, func(isFinished *bool) {
+		headerForMining.SetNonce(rand.Uint64())
+		pow.CheckProofOfWorkWithTarget(headerForMining, targetDifficulty)
+		hashes++
+	})
+
+	return machineHashesPerSecondMeasurementDuration.Nanoseconds() / hashes
+}
+
+func runDAATest(t *testing.T, testName string, runDuration time.Duration,
+	targetHashNanosecondsFunction func(totalElapsedTime time.Duration) int64) {
+
+	t.Logf("TEST STARTED: %s", testName)
+	defer t.Logf("TEST FINISHED: %s", testName)
+
+	tearDownKaspad := runKaspad(t)
+	defer tearDownKaspad()
+
+	rpcClient, err := rpcclient.NewRPCClient(rpcAddress)
+	if err != nil {
+		t.Fatalf("NewRPCClient: %s", err)
+	}
+
+	hashDurations := make([]time.Duration, 0, averageHashRateSampleSize+1)
+	miningDurations := make([]time.Duration, 0, averageBlockRateSampleSize+1)
+	previousDifficulty := float64(0)
+	blocksMined := 0
+
+	startTime := time.Now()
+	runForDuration(runDuration, func(isFinished *bool) {
+		getBlockTemplateResponse, err := rpcClient.GetBlockTemplate(miningAddress)
+		if err != nil {
+			t.Fatalf("GetBlockTemplate: %s", err)
+		}
+		templateBlock, err := appmessage.RPCBlockToDomainBlock(getBlockTemplateResponse.Block)
+		if err != nil {
+			t.Fatalf("RPCBlockToDomainBlock: %s", err)
+		}
+		targetDifficulty := difficulty.CompactToBig(templateBlock.Header.Bits())
+		headerForMining := templateBlock.Header.ToMutable()
+
+		miningStartTime := time.Now()
+		for i := rand.Uint64(); i < math.MaxUint64; i++ {
+			hashStartTime := time.Now()
+
+			headerForMining.SetNonce(i)
+			if pow.CheckProofOfWorkWithTarget(headerForMining, targetDifficulty) {
+				templateBlock.Header = headerForMining.ToImmutable()
+				break
+			}
+
+			// Yielding a thread in Go takes up to a few milliseconds whereas hashing once
+			// takes a few hundred nanoseconds, so we spin in place instead of e.g. calling time.Sleep()
+			for {
+				targetHashNanoseconds := targetHashNanosecondsFunction(time.Since(startTime))
+				hashElapsedTimeNanoseconds := time.Since(hashStartTime).Nanoseconds()
+				if hashElapsedTimeNanoseconds >= targetHashNanoseconds {
+					break
+				}
+			}
+
+			hashDuration := time.Since(hashStartTime)
+			hashDurations = append(hashDurations, hashDuration)
```

Comment:
Not sure if relevant or not, but this append might affect the durations of things, since it will be copying 100k items once it reaches capacity.

Reply:
It would, but note that if it reaches `averageHashRateSampleSize` the next couple of lines pop the oldest entry.
Also note that the capacity is defined as `averageHashRateSampleSize + 1`